dashboard/src/main/resources/templates/dashboard.ftl

```diff
@@ -21,9 +25,22 @@
       </tr>
       <#list table as row>
         <tr>
-          <td><a href='${row.getCoordinates()?replace(":", "_")}.html'>${row.getCoordinates()}</a></td>   
-          <td class='${row.getResult("Upper Bounds")?string('PASS', 'FAIL')}'>${row.getResult("Upper Bounds")?string('PASS', 'FAIL')}</td>
-          <td class='${row.getResult("Dependency Convergence")?string('PASS', 'FAIL')}'>${row.getResult("Dependency Convergence")?string('PASS', 'FAIL')}</td>
+          <#if row.getResult("Upper Bounds")?? ><#-- checking isNotNull() -->
+            <#-- When it's not null, it means the test ran. It's either PASS or FAIL -->
+            <#assign upper_bound_test_label = row.getResult("Upper Bounds")?then('PASS', 'FAIL') >
+          <#else>
+            <#-- Null means there's an exception and test couldn't run -->
+            <#assign upper_bound_test_label = "UNAVAILABLE">
+          </#if>
+          <td><a href='${row.getCoordinates()?replace(":", "_")}.html'>${row.getCoordinates()}</a></td>
+          <td class='${upper_bound_test_label}' title="${row.getExceptionMessage()!""}">${upper_bound_test_label}</td>
```

Comment:
double double quote at end of ${row.getExceptionMessage()!""}" looks suspicious

Reply:
Actually this doesn’t cause problem. Freemarker processes ${}, which contain only one pair of double quotes.

If you feel problem in readability, I’m happy to update this piece.

Reply:
Well, as you had to add a comment, readability is not good. I’ll use a pair of single quotes to easily distinguish them.

Reply:
Is this some weirdness with the exclamation point? E.g. if it's null then return the empty string instead or something like that?

Reply:
Exactly. If it's null, then empty string.